### ISSUE TITLE
Final TypeAlias incorrectly reported as "Variable not allowed in type expression"

### ISSUE THREAD

**Issue by dev-hanz-ops** [microsoft/pyright#8504]
**Describe the bug**

When defining a `Final[TypeAlias]` and using it as type hint, pyright reports: `error: Variable not allowed in type expression (reportInvalidTypeForm)`

**Code or Screenshots**

```python
class TypeTest:
    type_alias: Final[TypeAlias] = str
    test_list: list[type_alias] = ["a"]

    def modification_not_allowed(self):
        self.type_alias = float


type_test = TypeTest()
type_test.type_alias = int
TypeTest.type_alias = int
```

All three assignments are correctly reported as impossible:
```bash
... - error: Cannot assign to attribute "type_alias" for class "TypeTest*"
    Attribute "type_alias" cannot be assigned through a class instance because it is a ClassVar
    "type_alias" is declared as Final and cannot be reassigned (reportAttributeAccessIssue)
... - error: Cannot assign to attribute "type_alias" for class "TypeTest"
    Attribute "type_alias" cannot be assigned through a class instance because it is a ClassVar
    "type_alias" is declared as Final and cannot be reassigned (reportAttributeAccessIssue)
... - error: Cannot assign to attribute "type_alias" for class "type[TypeTest]"
    "type_alias" is declared as Final and cannot be reassigned (reportAttributeAccessIssue)

```

**Comment by dev-hanz-ops** [comment]
Similar case:

```python
class TypeTest:
    def __init__(self):
        type_alias: Final = str
        self.test_list: list[type_alias] = ["a"]
        type_alias = int
```

```bash
... - error: Variable not allowed in type expression (reportInvalidTypeForm)
... - error: "type_alias" is declared as Final and cannot be reassigned (reportGeneralTypeIssues)
```

**Comment by erictraut** [comment]
Pyright is working correctly here, so I don't consider this a bug.

You are using `TypeAlias` and `Final` incorrectly here. `TypeAlias` and `Final` are both special forms, and neither are allowed in a type expression. They are allowed in annotation expressions, but they cannot be combined. See [this section of the typing spec](https://typing.readthedocs.io/en/latest/spec/annotations.html#type-and-annotation-expressions) for details.

`Final` defines a variable that can be assigned a value only once. `TypeAlias` defines a type alias. It doesn't make sense to combine the two because variables are type aliases are very different concepts with different rules. Pyright treats the combination of the two as a variable, and variables are not allowed to be used in type expressions (as indicated by the error message).

In most languages, `TypeAlias` and `Final` would be keywords (like `type` and `const`). In fact, in Python 3.12 a new keyword `type` was added to replace the use of `TypeAlias`.

**Comment by dev-hanz-ops** [comment]
Hi @erictraut, thanks for providing the link to the relevant spec. I come to a different conclusion. In the light of below excerpts, can you explain why what I do is incorrect?

```
annotation_expression ::= ...
                           | <Final> '[' annotation_expression']'
                           | <TypeAlias>
                                 (valid only in variable annotations)
```

`TypeAlias` is an annotation expression. As per the doc, I can not chain an annotation expression after `TypeAlias`, but I can do so after `Final`. So `Final[TypeAlias]` is within spec. Furthermore:

> A type qualifier is a [special form](https://typing.readthedocs.io/en/latest/spec/glossary.html#term-special-form) that qualifies a [type expression](https://typing.readthedocs.io/en/latest/spec/glossary.html#term-type-expression) to form an [annotation expression](https://typing.readthedocs.io/en/latest/spec/glossary.html#term-annotation-expression). For example, the **type qualifier [Final](https://typing.readthedocs.io/en/latest/spec/qualifiers.html#uppercase-final) can be used around a type** to indicate that the annotated value may not be overridden or modified

[This last quote](https://typing.readthedocs.io/en/latest/spec/glossary.html#term-type-qualifier) says a) `Final` is a type qualifier and b) type qualifiers do qualify type expressions, directly contradicting

> `TypeAlias` and `Final` are both special forms, and neither are allowed in a type expression

And lastly, purely from a logical point of view:  The code defines a non-reassignable variable within a method, instead of globally. Why would that be not allowed, I can't find anything about `A TypeAlias can be defined only within a module or class scope (reportGeneralTypeIssues)` in the spec. Especially as reassigning a globally defined `TypeAlias` leads to a similar error as reassigning a `Final` variable:

```python
test: TypeAlias = str

class Test:
    def testmethod(self):
        global test
        test = int              # error: "test" is declared as a TypeAlias and can be assigned only once (reportRedeclaration)
        alias: TypeAlias = str  # error: A TypeAlias can be defined only within a module or class scope (reportGeneralTypeIssues)
        implicit_type = str     # no error
        impl: implicit_type     # error: Variable not allowed in type expression (reportInvalidTypeForm)
        final: Final = "a"
        final = "b"             # error: "final" is declared as Final and cannot be reassigned (reportGeneralTypeIssues)
        self.var: test = "abc"  # only error if 'global test' statement makes test a variable
```

**Comment by erictraut** [comment]
`Final[TypeAlias]` follows the grammar rules for an annotation expression, but it doesn't make sense semantically. Its meaning is nonsensical.

As I explained, `Final` defines a variable whereas `TypeAlias` defines a type alias. When a variable is assigned a value, the assigned expression is evaluated with _value expression_ rules. When a type alias is assigned a value, the assigned expression is evaluated with _type expression_ rules. Consider the following, where the exact same expression is assigned to two different symbols.

```python
A: Final = "Foo"
B: TypeAlias = "Foo"

class Foo: ...
```

`A` and `B` mean very different things here. `A` is a variable that has the type `Literal["Foo"]`. Since it's marked with a `Final` qualifier, it cannot be reassigned. `B` is a type alias that refers to the class `Foo`. The quotes in this case are interpreted as a forward reference, since type expression evaluation rules are used. 

It wouldn't make sense for a symbol to be interpreted as both a variable and a type alias. 

`Final` is a type qualifier in that it describes a non-type property of a symbol. In particular, it means that the symbol can be assigned a value only one time. 

A type alias can be defined only once. That makes sense because type aliases are used in type expressions, and type expressions are not always evaluated in code flow order. For example, if you place a type expression in quotes or use `from __future__ import annotated`, a type expression uses deferred evaluation. For this reason, it's important for a type alias to have only one unambiguous definition. 

As I mentioned before, it may help you to think about the concepts of `Final` and `TypeAlias` as keywords, as they would be in most programming languages. For example, TypeScript has the keywords `const` and `type` to declare constants and type aliases. 

```ts
const one = 1
type NumberOrString = number | string
```

It wouldn't make sense to have a "const type". Likewise, it doesn't make sense to have a `Final TypeAlias`.

**Comment by erictraut** [comment]
Let's walk through the four cases in your above example.

* `x: TypeAlias = int` is a type alias declaration. It uses the `TypeAlias` qualifier introduced in [PEP 613](https://peps.python.org/pep-0613/). This qualifier is outdated as of Python 3.12, which introduced a new preferred way to declare type aliases. [PEP 695](https://peps.python.org/pep-0695/) introduced a dedicated `type` keyword to the language. If you're using Python 3.12, I recommend switching to this form. The `TypeAlias` form will be deprecated in time. You can read more about `TypeAlias` in [this section of the typing spec](https://typing.readthedocs.io/en/latest/spec/aliases.html#typealias).
* `x = int` is not a type alias declaration in this context. It's a simple variable assignment. If this statement were located within the global scope, it would be considered a type alias declaration. However, this is an archaic way to define a type alias. It was introduced in PEP 484 and was superseded by `TypeAlias` in PEP 613, which was superseded by `type` in PEP 695. This form was replaced because its meaning is ambiguous, since there is no clear way for type checkers to differentiate between a variable assignment and a type alias declaration. Different type checkers use various heuristics to distinguish between these cases, but these heuristics have never been documented in detail. I recommend avoiding the use of this archaic form.
* `x: type[int] = int` is a variable declaration combined with an assignment. It declares that `x` is a variable that can accept any value whose type is [assignable to](url) `type[int]`. The assignment portion of the statement then assigns a value to this variable.
* `x: Type[int] = int` is the same as the previous case, but it uses the ([now deprecated](https://peps.python.org/pep-0585/)) `typing.Type` alias for `type`. If you're using Python 3.9 or newer, I recommend avoiding the use of these deprecated forms. They are unlikely to be removed from the typing module any time soon, but they are no longer preferred.

The typing spec is currently silent on whether PEP 613 and PEP 695 type alias declarations are allowed in function bodies. The only statement related to this can be found at the end of [this section](https://typing.readthedocs.io/en/latest/spec/aliases.html#typealias) where it clearly states that old-style (PEP 484) type alias declarations are allowed only in the global scope. The general consensus among type checkers is that PEP 613 and PEP 695 declarations should be allowed in both the global scope and in class scopes. This allows a class to act as a namespace for the type alias. The general consensus among type checker maintainers is that type aliases don't make sense within function bodies because function-level symbols are not visible outside of functions. However, this point has not yet been fully debated or codified in the typing spec. There is ongoing work to nail down underspecified parts of the typing spec, and this is one such issue. I don't plan to make any changes to pyright's behavior until this matter is resolved in the spec because any such change would cause churn for pyright users. I think it's likely that the spec will allow type alias declarations within function scopes (that is my preference, anyway), so I recommend that you write your code accordingly.

**Comment by dev-hanz-ops** [comment]
Many thanks for taking the time for these thorough explanations, very appreciated
